Public/Src/App/Bxl/AzureDevOpsListener.cs

```diff
@@ -0,0 +1,212 @@
+
+namespace BuildXL
+{
+    /// <summary>
+    /// This event listener should only be hooked up with AzureDevOps optimzied UI is requested
+    /// It tracks data that is not easily captured at the site where the information is emitted because
+    /// it is too hard to track through. It is most ideal to capture data at the site of collections.
+    /// But as an example the cache miss analyzer is many levels of indirection deep and passing
+    /// the BuildSummary model down would be way to cumbersome.
+    /// </summary>
+    public sealed class AzureDevOpsListener : FormattingEventListener
+    {
+        private static readonly char[] s_newLineCharArray = Environment.NewLine.ToCharArray();
+
+        private readonly IConsole m_console;
+
+        /// <summary>
+        /// The last reported percentage. To avoid double reporting the same percentage over and over
+        /// </summary>
+        private int m_lastReportedProgress = -1;
+
+        private readonly BuildViewModel m_buildViewModel;
+
+        /// <nodoc />
+        public AzureDevOpsListener(
+            Events eventSource,
+            IConsole console,
+            DateTime baseTime,
+            BuildViewModel buildViewModel)
+            : base(eventSource, baseTime, null, EventLevel.Verbose, false, TimeDisplay.Seconds)
+        {
+            Contract.Requires(eventSource != null);
+            Contract.Requires(console != null);
+            Contract.Requires(buildViewModel != null);
+
+            m_console = console;
+            m_buildViewModel = buildViewModel;
+        }
+
+
+        /// <inheritdoc />
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1063:ImplementIDisposableCorrectly")]
+        public override void Dispose()
+        {
+            m_console.Dispose();
+            base.Dispose();
+        }
+
+        /// <inheritdoc />
+        protected override void OnInformational(EventWrittenEventArgs eventData)
+        {
+            switch (eventData.EventId)
+            {
+                case (int)EventId.PipStatus:
+                case (int)BuildXL.Scheduler.Tracing.LogEventId.PipStatusNonOverwriteable:
+                    {
+                        var payload = eventData.Payload;
+
+                        var executing = (long)payload[10];
+                        var succeeded = (long)payload[11];
+                        var failed = (long)payload[12];
+                        var skipped = (long)payload[13];
+                        var pending = (long)payload[14];
+                        var waiting = (long)payload[15];
+
+                        var done = succeeded + failed + skipped;
+                        var total = done + pending + waiting + executing;
+
+                        var processPercent = (100.0 * done) / (total * 1.0);
+                        var currentProgress = Convert.ToInt32(Math.Floor(processPercent));
+
+                        if (m_lastReportedProgress != currentProgress)

```

Comment:
>if (m_lastReportedProgress != currentProgress) [](start = 24, length = 46)

I think it should be ` if (m_lastReportedProgress < currentProgress)`
scenario:
pip#1, very slow, high-priority pip
pip#2 - pip#K - pips that depend on pip#1

all other pips - are small/fast ones and do not depend on pip#1

with the current condition, the reported progress value will be oscillating between something small and big... I think the 'progress value' must be something that is continuously increasing.

Reply:
it is already rounded down to int, but I'll change.

---
In reply to: [316888774](https://github.com/microsoft/BuildXL/pull/784#discussion_r316888774) [](ancestors = 316888774)